File: src/main/java/progerbot/CodeHighlighter.java
package progerbot;

import java.io.*;
import java.util.List;
import java.util.ArrayList;
import java.util.Properties;

import gui.ava.html.image.generator.HtmlImageGenerator;
import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.message.BasicNameValuePair;


public class CodeHighlighter {
    private final String USER_AGENT = "Mozilla/5.0";

    private static String addStyles(String highlightedCode) {
        try {
            // loading token from locally stored file
            Properties prop = new Properties();
            InputStream inputStream =
                    MainServlet.class.getClassLoader().getResourceAsStream("code-highlight-styles.properties");
            prop.load(inputStream);
            String prefix = prop.getProperty("json.prefix");
            String suffix = prop.getProperty("json.suffix");
            return prefix + highlightedCode + suffix;

        } catch (IOException e) {
            e.printStackTrace();
            return "";
        }
    }

    // HTTP POST request
    private void sendPost() throws Exception {

        //String url = "https://selfsolve.apple.com/wcResults.do";
        String url = "http://pygments.simplabs.com/";

        org.apache.http.client.HttpClient client = new DefaultHttpClient();
        HttpPost post = new HttpPost(url);

        // add header
        post.setHeader("User-Agent", USER_AGENT);

        List<NameValuePair> urlParameters = new ArrayList<NameValuePair>();

        urlParameters.add(new BasicNameValuePair("lang", "java"));
        urlParameters.add(new BasicNameValuePair("code", "if (foo) then\n\tbar();\nelse\n\treturn;"));

        post.setEntity(new UrlEncodedFormEntity(urlParameters));

        HttpResponse response = client.execute(post);
        System.out.println("\nSending 'POST' request to URL : " + url);
        System.out.println("Post parameters : " + post.getEntity());
        System.out.println("Response Code : " +
                response.getStatusLine().getStatusCode());

        BufferedReader rd = new BufferedReader(
                new InputStreamReader(response.getEntity().getContent()));

        StringBuffer result = new StringBuffer();
        String line = "";
        while ((line = rd.readLine()) != null) {
            result.append(line);
            result.append("\n");
        }
        result.deleteCharAt(result.length() - 1);

        System.out.println(result.toString());

        HtmlImageGenerator imageGenerator = new HtmlImageGenerator();
        //imageGenerator.loadHtml("<b>Hello World!</b> Please goto <a title=\"Goto Google\" href=\"http://www.google.com\">Google</a>.");
        imageGenerator.loadHtml(addStyles(result.toString()));
        imageGenerator.saveAsImage("hello-world.png");

        //imageGenerator.saveAsHtmlWithMap("hello-world.html", "hello-world.png");

        //BufferedImage img = imageGenerator.getBufferedImage();
        //File f = new File("TEST_PNG_PICTURE.png");
        //ImageIO.write(img, "PNG", f);

/*

        BufferedImage img = null;
        try {
            img = ImageIO.read(new File("pic.png"));
        } catch (IOException e) {
            e.printStackTrace();
        }

        HttpPost post1 = new HttpPost("server_address");

        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        try {
            ImageIO.write(img, "png", baos);
        } catch (IOException e) {
            e.printStackTrace();
        }

        HttpEntity entity = MultipartEntityBuilder.create()
                .addBinaryBody("file_attach", baos.toByteArray()).build();

        post1.setEntity(entity);
        HttpClient client1 = HttpClientBuilder.create().build();
        HttpResponse response1 = null;
        try {
            response1 = client1.execute(post1);
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }*/
    }

    public static void main(String[] args) throws Exception {

        CodeHighlighter http = new CodeHighlighter();

        //System.out.println("Testing 1 - Send Http GET request");
        //http.sendGet();

        System.out.println("\nTesting 2 - Send Http POST request");
        http.sendPost();

    }
}

File: src/main/java/progerbot/CodeHighlighter.kt
package progerbot

import com.google.appengine.api.urlfetch.*
import com.google.gson.JsonParser
import gui.ava.html.image.generator.HtmlImageGenerator
import org.apache.http.entity.ContentType
import org.apache.http.entity.mime.MultipartEntityBuilder
import java.io.BufferedReader
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.StringReader
import java.net.URL
import java.util.*
import javax.imageio.ImageIO


public object CodeHighlighter {
    private val PYGMENTS_URL = "http://pygments.simplabs.com/"
    private val CHARSET = "UTF-8"
    /**
     * HTML prefix containing styles for highlighted text.
     */
    private val prefix : String
    /**
     * HTML suffix of a highlighted text.
     */
    private val suffix : String
    /**
     * A strategy for converting HTML to images.
     */
    private val imageObtainer : ImageObtainingStrategy = LibraryImageObtainingStrategy()

    init {
        try {
            // loading token from locally stored file
            val prop = Properties()
            val inputStream = MainServlet::class.java.
                    classLoader.getResourceAsStream("res.properties")
            prop.load(inputStream)
            prefix = prop.getProperty("json.highlightPrefix")
            suffix = prop.getProperty("json.highlightSuffix")

        }
        catch (e : IOException) {
            e.printStackTrace()
            prefix = ""
            suffix = ""
        }
    }

    /**
     * Adds styles (colors) to pygmentized code.
     */
    private fun addStyles(highlightedCode : String) = prefix + highlightedCode + suffix

    public fun manageCodeHighlightRequest(language : String, content : String, chatId : String, apiUrl : String) : Boolean {
        // creating and sending HTTP request to Pygments
        val pygmentsPost = HTTPRequest(URL(PYGMENTS_URL), HTTPMethod.POST)
        pygmentsPost.payload = "lang=$language&code=$content".toByteArray(CHARSET)
        val pygmentsResponse = URLFetchServiceFactory.getURLFetchService().fetch(pygmentsPost)
        if (pygmentsResponse.responseCode != 200) {
            Logger.println("[ERROR] Could not pygmentize code for chat $chatId")
            return false
        }
        // obtaining image with colored code
        val imageByteArray = imageObtainer.getImageFromHtml(
                addStyles(pygmentsResponse.content.toString()),
                apiUrl,
                chatId
        )
        // FIXME: epic workaround
        if (imageByteArray.size == 1) {
            return true
        }
        // creating Telegram POST request
        val telegramPost = HTTPRequest(URL(apiUrl + "/sendPhoto"), HTTPMethod.POST)
        val entity = (MultipartEntityBuilder.create()
                .addTextBody("chat_id", chatId)
                .addBinaryBody("photo", imageByteArray, ContentType.MULTIPART_FORM_DATA, "image.png"))
                .build()
        val bos = ByteArrayOutputStream()
        entity.writeTo(bos)
        val body = bos.toByteArray()
        // extract multipart boundary (body starts with --boundary\r\n)
        var boundary = BufferedReader(StringReader(String(body))).readLine()
        boundary = boundary.substring(2, boundary.length)
        // adding multipart header and body
        telegramPost.addHeader(HTTPHeader("Content-type", "multipart/form-data; boundary=" + boundary))
        telegramPost.payload = body
        // sending image back to user
        val telegramResponse = URLFetchServiceFactory.getURLFetchService().fetch(pygmentsPost)
        return telegramResponse.responseCode == 200
    }

    private abstract class ImageObtainingStrategy() {
        public abstract fun getImageFromHtml(htmlContent : String, telegramApiUrl : String, chatId : String) : ByteArray
    }

    private class LibraryImageObtainingStrategy() : ImageObtainingStrategy() {
        override fun getImageFromHtml(htmlContent : String, telegramApiUrl : String, chatId : String) : ByteArray {
            val imageGenerator = HtmlImageGenerator()
            imageGenerator.loadHtml(htmlContent)
            val baos = ByteArrayOutputStream()
            ImageIO.write(imageGenerator.bufferedImage, "png", baos)
            return baos.toByteArray()
        }
    }

    private class ApiImageObtainingStrategy() : ImageObtainingStrategy() {
        private val API_URL = "http://api.page2images.com/html2image"
        private val apiKey : String

        init {
            // loading properties
            val prop = Properties()
            val inputStream = progerbot.MainServlet::class.java.
                    classLoader.getResourceAsStream("auth.properties")
            prop.load(inputStream)
            apiKey = prop.getProperty("json.pagetoimagesApiKey")
        }

        override fun getImageFromHtml(htmlContent : String, telegramApiUrl : String, chatId : String) : ByteArray {
            val response : HTTPResponse = HttpRequests.simpleRequest(
                    API_URL,
                    HTTPMethod.POST,
                    "p2i_html=$htmlContent&p2i_key=$apiKey&p2i_imageformat=jpg&p2i_fullpage=0"
            )
            val parser = JsonParser()
            val jsonObject = parser.parse(response.content.toString()).asJsonObject
            when (jsonObject.get("status").asString) {
                "processing" -> {
                    HttpRequests.simpleRequest(
                            "$telegramApiUrl/sendMessage",
                            HTTPMethod.POST,
                            "chat_id=$chatId&text=Your photo is being prepared now. But I won't give it to you, haha."
                    )
                }
                "finished" -> {
                    HttpRequests.simpleRequest(
                            "$telegramApiUrl/sendMessage",
                            HTTPMethod.POST,
                            "chat_id=$chatId&text=Here you are:\n${jsonObject.get("image_url").asString}"
                    )
                }

            }
            return ByteArray(1)
        }
    }
}